core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -432,6 +465,93 @@ private[spark] class Executor(
   }
 
   /**
+   * Supervises the killing / cancellation of a task by sending the interrupted flag, optionally
+   * sending a Thread.interrupt(), and monitoring the task until it finishes.
+   */
+  private class TaskReaper(
```

Comment:
It will be good to explain more about how this class works. It is great that we have comments at specific parts, but a section explains the workflow will be very helpful to future readers.